feat(curriculum):  add Data Structures Quiz (#56439)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/quiz-javascript-data-structures/66edd5583761e4f6dae19606.md

#### --answer--

Last in, first out (LIFO)

### --question--

#### --text--

A set data structure: 

#### --distractors--

Guarantees sorted order

---

Allows duplicate elements

---

Allows access by index

#### --answer--

Does not allow duplicate elements

### --question--

#### --text--

What differentiates a dynamic array from a static array?

#### --distractors--

Dynamic arrays require less memory than Static Arrays

---

Static arrays are always slower

---

Static arrays are only used in Java

#### --answer--

Dynamic arrays can grow whereas static arrays have fixed size 

### --question--

#### --text--

What is the time complexity of inserting an element in a binary heap?

#### --distractors--

O(1)

---

O(n)

---

O(n log n)

#### --answer--

O(log n)

### --question--

#### --text--

In a circular queue, if rear == front, what does it indicate?

#### --distractors--

Queue is at capacity

---

Queue is full

---

Queue is undefined

#### --answer--

Queue is empty

### --question--

#### --text--

Which of the following is a disadvantage of static arrays?

#### --distractors--

Expensive insertion

---

Expensive access time

---

Unsorted Elements

#### --answer--

> Fixed Size

Reviewer: ```suggestion
Fixed Size

```

you have accidentally removed a line here

---

Less than or equal to the value of its children

#### --answer--

Greater than or equal to the value of its children

### --question--

#### --text--

Which of the following is true for a singly linked list?

#### --distractors--

It allows traversal in both directions

---

Each node has two pointers

---

The list is circular by default

#### --answer--

Each node contains data and a pointer to the next node

### --question--

#### --text--

What is the primary benefit of using a hash set over an array?

#### --distractors--

Reduced memory usage

---

Random access to elements

---

Maintaining order of elements

#### --answer--

Faster insertion and search times due to hash function

### --question--

#### --text--

Which operation is faster in a doubly linked list compared to a singly linked list?

#### --distractors--

Searching for an element

---

Reversing the list

---

Insertion at the end

#### --answer--

Deleting a node when a reference to that node is provided

### --question--

#### --text--

What is the main difference between a queue and a stack?

#### --distractors--

Stack allows random access while queue does not

---

> Queue follows LIFO and stack follows FIFO

Reviewer 2: ```suggestion
Queue follows "Last in, first out" (LIFO) and stack follows "First in, first out" (FIFO)
```

---

Both follow the same order of operations

#### --answer--

> Stack follows LIFO and queue follows FIFO

Reviewer 2: ```suggestion
Stack follows "Last in, first out" (LIFO) and queue follows "First in, first out" (FIFO)
```

#### --text--

Which of the following operations is not performed on a stack?

#### --distractors--

Push

---

Pop

---

Peek

#### --answer--

Enqueue

### --question--

#### --text--

What happens to the time complexity of appending an element to a dynamic array when it is resized due to capacity limits?

#### --distractors--

Always O(1)

---

Always O(n)

---

O(log n)

#### --answer--

O(1) amortized, but O(n) when resizing occurs

### --question--

#### --text--

> Which of the following is a property of a binary min-heap?
> 

Reviewer 3: ```suggestion
Which of the following is a property of a binary min-heap?
```

Extra line break

#### --distractors--

Tree is always balanced

---

Root contains the largest element

---

Elements are sorted in ascending order

#### --answer--

Root contains the smallest element

### --question--

#### --text--

> Choose the coorect statement about hash collisions in a hash map?

Reviewer 3: ```suggestion
Choose the correct statement about hash collisions in a hash map.
```

Not a question, so should end with a full-stop. Also, a typo.

#### --distractors--

They are resolved using sorting

---

They are impossible to occur

---

They can crash the program

#### --answer--

They are resolved using linked lists (chaining method) or open addressing

### --question--

#### --text--

What is the time complexity of inserting a node at the end of a singly linked list?

#### --distractors--

O(1)

---

O(log n)

---

O(n log n)

#### --answer--

O(n)

### --question--

#### --text--

What is the time complexity of accessing an element in a hash map in the average case?

#### --distractors--

O(n)

---

O(n log n)

---

O(log n)

#### --answer--

O(1)

### --question--

#### --text--

Which of the following is used to implement a priority queue?

#### --distractors--

Stack

---

Queue

---

Linked List

#### --answer--

Heap

### --question--

#### --text--

> In a doubly linked list, each node contains ?

Reviewer 3: ```suggestion
In a doubly linked list, what does each node contain?
```

Grammar nit.

#### --distractors--

Data and only a pointer to the previous node

---

Data and a pointer to the next node

---

Data only

#### --answer--

Data and two pointers (next and previous)

### --question--

#### --text--

> The value of each node in a max-heap is?

Reviewer 3: ```suggestion
The value of each node in a max-heap is:
```

#### --distractors--

Equal to its children

---

Unrelated to the value of its children

---

Less than or equal to the value of its children

#### --answer--

Greater than or equal to the value of its children

### --question--

#### --text--

Which of the following is true for a singly linked list?

#### --distractors--

It allows traversal in both directions

---

Each node has two pointers

---

> The list is circular by default
> 

Reviewer 3: ```suggestion
The list is circular by default
```

Extra new line.

#### --answer--

Each node contains data and a pointer to the next node

### --question--

#### --text--

What is the primary benefit of using a hash set over an array?

#### --distractors--

Reduced memory usage

---

Random access to elements

---

Maintaining order of elements

#### --answer--

Faster insertion and search times due to hash function

### --question--

#### --text--

Which operation is faster in a doubly linked list compared to a singly linked list?

#### --distractors--

Searching for an element

---

Reversing the list

---

Insertion at the end

#### --answer--

Deleting a node when a reference to that node is provided

### --question--

#### --text--

What is the main difference between a queue and a stack?

#### --distractors--

Stack allows random access while queue does not

---

Queue follows LIFO and stack follows FIFO

---

Both follow the same order of operations

#### --answer--

Stack follows LIFO and queue follows FIFO

### --question--

#### --text--

What is the time complexity of accessing an element in a static array?

#### --distractors--

> O( log n)

Reviewer 3: ```suggestion
O(log n)
```

---

O(n log n)

---

O(n)

#### --answer--

O(1)

### --question--

#### --text--

What is the order of insertion and deletion in a stack?

#### --distractors--

Random access

---

First in, last out (FILO)

---

First in, first out (FIFO)

#### --answer--

Last in, first out (LIFO)

### --question--

#### --text--

A set data structure: 

#### --distractors--

Guarantees sorted order

---

Allows duplicate elements

---

Allows access by index

#### --answer--

Does not allow duplicate elements

### --question--

#### --text--

> What differs a dynamic array from a static array?

Reviewer 3: ```suggestion
What differentiates a dynamic array from a static array?
```

#### --distractors--

Dynamic arrays require less memory than Static Arrays

---

Static arrays are always slower

---

Static arrays are only used in Java

#### --answer--

> Dynamic arrays can grow whereas Static arrays have fixed size 

Reviewer 3: ```suggestion
Dynamic arrays can grow whereas static arrays have fixed size 
```

Author: Hey!! A big thanks for giving me guidance and sorry for the mistakes!!
